packages/jsii-pacmak/lib/targets/java.ts

```diff
@@ -1513,12 +1522,211 @@ class JavaGenerator extends Generator {
     if (method.abstract && !defaultImpl) {
       this.code.line(`${modifiers.join(' ')} ${signature};`);
     } else {
-      this.code.openBlock(`${modifiers.join(' ')} ${signature}`);
+      this.code.openBlock(
+        `${modifiers.join(' ')} ${signature}`,
+      );
+      this.emitUnionParameterValdation(method.parameters);
       this.code.line(this.renderMethodCall(cls, method, async));
       this.code.closeBlock();
     }
   }
 
+  /**
+   * Emits type checks for values passed for type union parameters.
+   *
+   * @param parameters the list of parameters received by the function.
+   */
+  private emitUnionParameterValdation(
+    parameters?: readonly spec.Parameter[],
+  ): void {
+    if (!this.runtimeTypeChecking) {
+      // We were configured not to emit those, so bail out now.
+      return;
+    }
+    const unionParameters = parameters?.filter(({ type }) =>
+      containsUnionType(type),
+    );
+    if (unionParameters == null || unionParameters.length === 0) {
+      return;
+    }
+
+    this.code.openBlock('if (software.amazon.jsii.Configuration.getRuntimeTypeChecking())');
+    for (const param of unionParameters) {
+      if (param.variadic) {
+        const javaType = this.toJavaType(param.type);
+        const asListName = `__${param.name}__asList`;
+        this.code.line(
+          `java.util.ArrayList<${javaType}> ${asListName} = new java.util.ArrayList<${javaType}>(java.util.Arrays.asList(${param.name}));`,
```

Comment:
We can't do:

```
java.util.Arrays.<${javaType}>asList(${param.name});`,
```

because the `asList()` returns a fixed size list (meaning that subsequent `add()` operations will throw `UnsupportedOperationException()`.